File: example-usage/use-devdigger-db.js
#!/usr/bin/env node

/**
 * Example: How to use DevDigger database in another project
 * 
 * This script shows how to:
 * 1. Connect to the DevDigger SQLite database
 * 2. Query crawled data
 * 3. Perform semantic search
 * 4. Export data for use in other applications
 */

const Database = require('better-sqlite3');
const path = require('path');
const os = require('os');

// Path to DevDigger database
const DB_PATH = path.join(
  os.homedir(),
  'Library',
  'Application Support',
  'devdigger',
  'devdigger.db'
);

console.log('📁 Connecting to DevDigger database at:', DB_PATH);

// Connect to the database
const db = new Database(DB_PATH, { readonly: true });

// ============================================
// 1. LIST ALL SOURCES (CRAWLED WEBSITES)
// ============================================
function listSources() {
  console.log('\n📚 SOURCES IN DATABASE:');
  console.log('=' .repeat(50));
  
  const sources = db.prepare(`
    SELECT 
      id,
      type,
      url,
      title,
      crawl_status,
      datetime(created_at) as created_at
    FROM sources
    ORDER BY created_at DESC
  `).all();
  
  if (sources.length === 0) {
    console.log('No sources found. Please crawl some websites first in DevDigger.');
    return [];
  }
  
  sources.forEach(source => {
    console.log(`
📌 ${source.title || 'Untitled'}
   URL: ${source.url}
   Type: ${source.type}
   Status: ${source.crawl_status}
   ID: ${source.id}
   Created: ${source.created_at}
    `);
  });
  
  return sources;
}

// ============================================
// 2. GET DOCUMENTS FOR A SOURCE
// ============================================
function getDocuments(sourceId) {
  console.log('\n📄 DOCUMENTS:');
  console.log('=' .repeat(50));
  
  const documents = db.prepare(`
    SELECT 
      id,
      source_id,
      substr(content, 1, 200) as preview,
      chunk_index,
      length(content) as content_length
    FROM documents
    WHERE source_id = ?
    ORDER BY chunk_index
  `).all(sourceId);
  
  console.log(`Found ${documents.length} document chunks`);
  
  return documents;
}

// ============================================
// 3. SEARCH FOR CONTENT
// ============================================
function searchContent(query) {
  console.log(`\n🔍 SEARCHING FOR: "${query}"`);
  console.log('=' .repeat(50));
  
  const results = db.prepare(`
    SELECT 
      d.id,
      d.content,
      d.source_id,
      s.url,
      s.title
    FROM documents d
    JOIN sources s ON d.source_id = s.id
    WHERE d.content LIKE ?
    LIMIT 10
  `).all(`%${query}%`);
  
  if (results.length === 0) {
    console.log('No results found.');
    return [];
  }
  
  results.forEach((result, index) => {
    const preview = result.content.substring(0, 200).replace(/\n/g, ' ');
    console.log(`
Result ${index + 1}:
  Source: ${result.title || result.url}
  Preview: ${preview}...
    `);
  });
  
  return results;
}

// ============================================
// 4. GET CODE EXAMPLES
// ============================================
function getCodeExamples(language = null) {
  console.log('\n💻 CODE EXAMPLES:');
  console.log('=' .repeat(50));
  
  let query = `
    SELECT 
      c.id,
      c.language,
      c.code,
      c.description,
      s.url as source_url
    FROM code_examples c
    JOIN sources s ON c.source_id = s.id
  `;
  
  if (language) {
    query += ` WHERE c.language = ?`;
  }
  
  query += ` LIMIT 10`;
  
  const stmt = db.prepare(query);
  const examples = language ? stmt.all(language) : stmt.all();
  
  if (examples.length === 0) {
    console.log('No code examples found.');
    return [];
  }
  
  examples.forEach(example => {
    console.log(`
Language: ${example.language}
Description: ${example.description || 'No description'}
Source: ${example.source_url}
Code Preview:
${example.code.substring(0, 200)}...
    `);
  });
  
  return examples;
}

// ============================================
// 5. EXPORT DATA AS JSON
// ============================================
function exportToJSON() {
  console.log('\n📦 EXPORTING DATA TO JSON...');
  console.log('=' .repeat(50));
  
  const data = {
    sources: db.prepare('SELECT * FROM sources').all(),
    documents: db.prepare('SELECT * FROM documents').all(),
    code_examples: db.prepare('SELECT * FROM code_examples').all(),
    collections: db.prepare('SELECT * FROM collections').all()
  };
  
  const fs = require('fs');
  const outputPath = './devdigger-export.json';
  
  fs.writeFileSync(outputPath, JSON.stringify(data, null, 2));
  console.log(`✅ Data exported to: ${outputPath}`);
  console.log(`   Sources: ${data.sources.length}`);
  console.log(`   Documents: ${data.documents.length}`);
  console.log(`   Code Examples: ${data.code_examples.length}`);
  
  return data;
}

// ============================================
// 6. USE WITH VECTOR SEARCH (EMBEDDINGS)
// ============================================
function vectorSearch(embedding) {
  console.log('\n🧠 VECTOR SEARCH:');
  console.log('=' .repeat(50));
  
  // This requires the embedding column to have data
  const results = db.prepare(`
    SELECT 
      d.id,
      d.content,
      s.url,
      s.title,
      cosine_similarity(d.embedding, ?) as similarity
    FROM documents d
    JOIN sources s ON d.source_id = s.id
    WHERE d.embedding IS NOT NULL
    ORDER BY similarity DESC
    LIMIT 10
  `).all(Buffer.from(new Float32Array(embedding).buffer));
  
  return results;
}

// ============================================
// 7. GET DATABASE STATISTICS
// ============================================
function getStats() {
  console.log('\n📊 DATABASE STATISTICS:');
  console.log('=' .repeat(50));
  
  const stats = {
    sources: db.prepare('SELECT COUNT(*) as count FROM sources').get().count,
    documents: db.prepare('SELECT COUNT(*) as count FROM documents').get().count,
    code_examples: db.prepare('SELECT COUNT(*) as count FROM code_examples').get().count,
    collections: db.prepare('SELECT COUNT(*) as count FROM collections').get().count,
    embedded_docs: db.prepare('SELECT COUNT(*) as count FROM documents WHERE embedding IS NOT NULL').get().count
  };
  
  console.log(`
  Sources: ${stats.sources}
  Documents: ${stats.documents}
  Code Examples: ${stats.code_examples}
  Collections: ${stats.collections}
  Documents with Embeddings: ${stats.embedded_docs}
  `);
  
  return stats;
}

// ============================================
// MAIN EXECUTION
// ============================================
async function main() {
  try {
    // 1. Get statistics
    const stats = getStats();
    
    if (stats.sources === 0) {
      console.log('\n⚠️  No data found in database!');
      console.log('Please use DevDigger to crawl some websites first.');
      console.log('\nTo crawl data:');
      console.log('1. Open DevDigger app');
      console.log('2. Go to "Mine" tab');
      console.log('3. Enter a URL (e.g., https://react.dev/learn)');
      console.log('4. Click "Commence"');
      console.log('5. Wait for crawling to complete');
      process.exit(0);
    }
    
    // 2. List sources
    const sources = listSources();
    
    // 3. Get documents for first source
    if (sources.length > 0) {
      getDocuments(sources[0].id);
    }
    
    // 4. Search for content
    searchContent('react');
    
    // 5. Get code examples
    getCodeExamples();
    
    // 6. Export data (optional)
    // exportToJSON();
    
  } catch (error) {
    console.error('Error:', error);
  } finally {
    db.close();
  }
}

// Run the script
main();

File: example-usage/use_devdigger_db.py
#!/usr/bin/env python3

"""
Example: How to use DevDigger database in Python projects

This script shows how to:
1. Connect to the DevDigger SQLite database
2. Query crawled data
3. Perform semantic search
4. Use with LangChain or other AI tools
"""

import sqlite3
import json
import os
from pathlib import Path
import numpy as np

# Path to DevDigger database
DB_PATH = Path.home() / "Library" / "Application Support" / "devdigger" / "devdigger.db"

class DevDiggerDB:
    """Interface to DevDigger knowledge database"""
    
    def __init__(self, db_path=DB_PATH):
        self.db_path = db_path
        self.conn = sqlite3.connect(str(db_path))
        self.conn.row_factory = sqlite3.Row
        self.cursor = self.conn.cursor()
        
    def get_stats(self):
        """Get database statistics"""
        stats = {}
        tables = ['sources', 'documents', 'code_examples', 'collections']
        
        for table in tables:
            self.cursor.execute(f"SELECT COUNT(*) as count FROM {table}")
            stats[table] = self.cursor.fetchone()['count']
            
        return stats
    
    def list_sources(self):
        """List all crawled sources"""
        self.cursor.execute("""
            SELECT id, type, url, title, crawl_status, created_at
            FROM sources
            ORDER BY created_at DESC
        """)
        return [dict(row) for row in self.cursor.fetchall()]
    
    def search(self, query, limit=10):
        """Full-text search across documents"""
        self.cursor.execute("""
            SELECT 
                d.id,
                d.content,
                d.source_id,
                s.url,
                s.title
            FROM documents d
            JOIN sources s ON d.source_id = s.id
            WHERE d.content LIKE ?
            LIMIT ?
        """, (f"%{query}%", limit))
        return [dict(row) for row in self.cursor.fetchall()]
    
    def get_documents(self, source_id=None):
        """Get all documents, optionally filtered by source"""
        if source_id:
            self.cursor.execute("""
                SELECT * FROM documents 
                WHERE source_id = ?
                ORDER BY chunk_index
            """, (source_id,))
        else:
            self.cursor.execute("SELECT * FROM documents")
        return [dict(row) for row in self.cursor.fetchall()]
    
    def get_code_examples(self, language=None):
        """Get code examples, optionally filtered by language"""
        if language:
            self.cursor.execute("""
                SELECT c.*, s.url as source_url
                FROM code_examples c
                JOIN sources s ON c.source_id = s.id
                WHERE c.language = ?
            """, (language,))
        else:
            self.cursor.execute("""
                SELECT c.*, s.url as source_url
                FROM code_examples c
                JOIN sources s ON c.source_id = s.id
            """)
        return [dict(row) for row in self.cursor.fetchall()]
    
    def get_embeddings(self):
        """Get documents with embeddings for vector search"""
        self.cursor.execute("""
            SELECT 
                d.id,
                d.content,
                d.embedding,
                s.url,
                s.title
            FROM documents d
            JOIN sources s ON d.source_id = s.id
            WHERE d.embedding IS NOT NULL
        """)
        
        results = []
        for row in self.cursor.fetchall():
            doc = dict(row)
            # Convert embedding from blob to numpy array
            if doc['embedding']:
                doc['embedding'] = np.frombuffer(doc['embedding'], dtype=np.float32)
            results.append(doc)
        return results
    
    def export_to_json(self, output_path="devdigger_export.json"):
        """Export entire database to JSON"""
        data = {
            'sources': self.list_sources(),
            'documents': self.get_documents(),
            'code_examples': self.get_code_examples()
        }
        
        with open(output_path, 'w') as f:
            json.dump(data, f, indent=2, default=str)
        
        return output_path
    
    def to_langchain_documents(self):
        """Convert to LangChain Document format"""
        try:
            from langchain.schema import Document
        except ImportError:
            print("Install langchain: pip install langchain")
            return []
        
        documents = []
        for doc in self.get_documents():
            documents.append(Document(
                page_content=doc['content'],
                metadata={
                    'source_id': doc['source_id'],
                    'chunk_index': doc['chunk_index'],
                    'id': doc['id']
                }
            ))
        return documents
    
    def close(self):
        """Close database connection"""
        self.conn.close()


# ============================================
# USAGE EXAMPLES
# ============================================

def example_basic_usage():
    """Basic usage example"""
    print("=" * 60)
    print("DEVDIGGER DATABASE USAGE EXAMPLE")
    print("=" * 60)
    
    # Connect to database
    db = DevDiggerDB()
    
    # Get statistics
    stats = db.get_stats()
    print(f"\n📊 Database Statistics:")
    for table, count in stats.items():
        print(f"  {table}: {count}")
    
    if stats['sources'] == 0:
        print("\n⚠️  No data found! Please crawl some websites in DevDigger first.")
        return
    
    # List sources
    print(f"\n📚 Sources:")
    for source in db.list_sources()[:5]:
        print(f"  - {source['title'] or source['url']}")
    
    # Search for content
    print(f"\n🔍 Searching for 'react':")
    results = db.search("react", limit=3)
    for i, result in enumerate(results, 1):
        preview = result['content'][:100].replace('\n', ' ')
        print(f"  {i}. {preview}...")
    
    # Get code examples
    print(f"\n💻 Code Examples:")
    examples = db.get_code_examples()[:3]
    for example in examples:
        print(f"  - {example['language']}: {example.get('description', 'No description')}")
    
    db.close()


def example_with_openai():
    """Example using with OpenAI for RAG"""
    from openai import OpenAI
    
    # Initialize
    client = OpenAI()  # Uses OPENAI_API_KEY env var
    db = DevDiggerDB()
    
    # Get relevant documents
    query = "How do React hooks work?"
    docs = db.search(query, limit=5)
    
    # Create context from documents
    context = "\n\n".join([doc['content'] for doc in docs])
    
    # Generate response with context
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "Answer based on the provided context."},
            {"role": "user", "content": f"Context:\n{context}\n\nQuestion: {query}"}
        ]
    )
    
    print(f"Answer: {response.choices[0].message.content}")
    db.close()


def example_with_langchain():
    """Example using with LangChain"""
    from langchain.vectorstores import Chroma
    from langchain.embeddings import OpenAIEmbeddings
    from langchain.chains import RetrievalQA
    from langchain.llms import OpenAI
    
    # Get documents from DevDigger
    db = DevDiggerDB()
    documents = db.to_langchain_documents()
    
    # Create vector store
    embeddings = OpenAIEmbeddings()
    vectorstore = Chroma.from_documents(documents, embeddings)
    
    # Create QA chain
    qa = RetrievalQA.from_chain_type(
        llm=OpenAI(),
        chain_type="stuff",
        retriever=vectorstore.as_retriever()
    )
    
    # Ask questions
    result = qa.run("What is useState in React?")
    print(result)
    
    db.close()


if __name__ == "__main__":
    # Check if database exists
    if not DB_PATH.exists():
        print(f"❌ Database not found at: {DB_PATH}")
        print("\nPlease use DevDigger to crawl some data first!")
        exit(1)
    
    # Run basic example
    example_basic_usage()
    
    # Uncomment to run advanced examples:
    # example_with_openai()
    # example_with_langchain()